pkg/operator/clustermembercontroller/clustermembercontroller.go

```diff
@@ -89,87 +107,586 @@ func (c *ClusterMemberController) reconcileMembers(ctx context.Context, recorder
+	memberMap := getMemberMap(members)
+	memberToMachineMap := make(map[string]memberAndMachine)
+
+	for _, machine := range machines {
+		// Get etcd member peerURL for this Machine
+		peerURL, err := c.getPeerURLForMachine(machine)
+		if err != nil {
+			if errors.Is(err, ErrMissingNodeOnMachine) {
+				// TODO(haseeb): Should this be a warning or silent?
+				klog.V(4).Infof("Can't map Machine (%s/%s) to etcd member. No node name found on status.NodeRef", machine.Namespace, machine.Name)
+				continue
+			}
+			return nil, fmt.Errorf("failed to get peerURL for machine: %w", err)
+		}
+		member, isMember := memberMap[peerURL]
+
+		// Ignore the machine if its node is not part of the current etcd cluster membership
+		if !isMember {
+			continue
+		}
+		memberToMachineMap[peerURL] = memberAndMachine{
+			member:  *member,
+			machine: machine,
+		}
+	}
+	return memberToMachineMap, nil
+}
+
+// getNodeToMachineMap returns a map of Node name to the Machine referencing it in its machine.Status.NodeRef
+func getNodeToMachineMap(machines []machinev1beta1.Machine) map[string]machinev1beta1.Machine {
+	nodeToMachineMap := make(map[string]machinev1beta1.Machine)
+	for _, machine := range machines {
+		if machine.Status.NodeRef == nil || machine.Status.NodeRef.Name == "" {
+			// TODO(haseeb): Should this be a warning or silent?
+			klog.V(4).Infof("Can't map Machine (%s/%s) to Node. No node name found on status.NodeRef", machine.Namespace, machine.Name)
+			continue
+		}
+		nodeToMachineMap[machine.Status.NodeRef.Name] = machine
+	}
+	return nodeToMachineMap
+}
+
+// getMemberMap returns a mapping of etcd member peerURL to member
+func getMemberMap(members []*etcdserverpb.Member) map[string]*etcdserverpb.Member {
+	memberMap := make(map[string]*etcdserverpb.Member)
+	for _, m := range members {
+		// The reason we use the PeerURL and not the member name is because
+		// the member.Name field is only populated once the member starts.
+		// So unstarted members will all have member.Name=""
+		memberMap[m.PeerURLs[0]] = m
+	}
+	return memberMap
+}
+
+func countVotingMembers(members []*etcdserverpb.Member) int {
+	numVoting := 0
+	for _, m := range members {
+		if !m.IsLearner {
+			numVoting++
+		}
+	}
+	return numVoting
+}
+
+func getPreDrainHook() machinev1beta1.LifecycleHook {
+	return machinev1beta1.LifecycleHook{
+		Name:  "EtcdQuorum",
+		Owner: "etcd-cluster-operator",
+	}
+}
+
+func hasPreDrainHook(hooks []machinev1beta1.LifecycleHook) bool {
+	preDrainHook := getPreDrainHook()
+	for _, h := range hooks {
+		if h == preDrainHook {
+			return true
+		}
+	}
+	return false
+}
+
+func filterPreDrainHook(hooks []machinev1beta1.LifecycleHook) []machinev1beta1.LifecycleHook {
+	preDrainHook := getPreDrainHook()
+	var filtered []machinev1beta1.LifecycleHook
+	for _, h := range hooks {
+		if h == preDrainHook {
+			continue
+		}
+		filtered = append(filtered, h)
+	}
+	return filtered
 }
```

Comment:
please split the machine hook into its own controller. The size of this controller is conerning. It does too much.